File: skills/wechat-draft-proxy/scripts/create_draft.py
#!/usr/bin/env python3
"""
微信公众号草稿箱创建工具（代理服务器版）
通过固定 IP 的代理服务器中转，无需在本机配置微信 IP 白名单。

配置优先级:
  1. 命令行 --server / --api-key
  2. 技能目录 .env 文件（cp .env.example .env 后编辑）
  3. 环境变量 WECHAT_PROXY_URL / WECHAT_PROXY_API_KEY

用法:
  python3 create_draft.py --title "标题" --file article.html --html
  python3 create_draft.py --title "标题" --file article.html --html --cover-path cover.jpg

作为标准 skill 使用时推荐从 skills/wechat-draft-proxy/ 目录调用（或安装到 ~/.grok/skills 或 ~/.claude/skills）。
"""

import argparse
import base64
import json
import os
import re
import sys

try:
    import requests
except ImportError:
    print("需要安装依赖: pip install -r requirements.txt (或 pip install requests)")
    sys.exit(1)


# ==================== .env 自动加载（零额外依赖，匹配 content-factory 风格） ====================
def _load_env_manual(env_path: str) -> None:
    """极简 .env 解析器（当没有 python-dotenv 时使用）"""
    if not os.path.exists(env_path):
        return
    try:
        with open(env_path, "r", encoding="utf-8") as f:
            for raw_line in f:
                line = raw_line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" not in line:
                    continue
                key, value = line.split("=", 1)
                key = key.strip()
                value = value.strip().strip('"').strip("'")
                # 不覆盖已存在的环境变量（shell export 优先）
                if key and key not in os.environ:
                    os.environ[key] = value
    except Exception:
        pass


def load_proxy_env() -> None:
    """从 wechat-draft-proxy 根目录的 .env 加载配置（仅填充未设置的变量）"""
    # scripts/ 的父目录即 skill 根目录
    skill_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    env_file = os.path.join(skill_root, ".env")
    if not os.path.exists(env_file):
        return

    # 优先尝试 python-dotenv（用户可选安装）
    try:
        from dotenv import load_dotenv  # type: ignore
        load_dotenv(dotenv_path=env_file, override=False)
    except ImportError:
        _load_env_manual(env_file)


# 导入时自动加载（最常见直接运行场景）
load_proxy_env()


# wechat-styler 路径（用于 Markdown → HTML 转换）
_STYLER_SCRIPTS_DIR = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
    "wechat-styler", "scripts"
)
_STYLER_THEMES_DIR = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
    "wechat-styler", "themes"
)


def markdown_to_html(markdown_text: str, theme: str = "professional-clean") -> str:
    if os.path.isdir(_STYLER_SCRIPTS_DIR):
        try:
            if _STYLER_SCRIPTS_DIR not in sys.path:
                sys.path.insert(0, _STYLER_SCRIPTS_DIR)
            from style_html import WeChatConverter, load_theme  # type: ignore
            converter = WeChatConverter(load_theme(theme, _STYLER_THEMES_DIR))
            result = converter.convert(markdown_text)
            print(f"  已使用 wechat-styler 排版（主题：{theme}）")
            return result.html
        except Exception as e:
            print(f"  wechat-styler 调用失败（{e}），fallback 到内置转换")

    # 内置简单转换（仅兜底）
    html = markdown_text
    for i in range(6, 0, -1):
        html = re.sub(rf'^#{i}\s+(.+)$', rf'<h{i}>\1</h{i}>', html, flags=re.MULTILINE)
    html = re.sub(r'\*\*(.+?)\*\*', r'<strong>\1</strong>', html)
    html = re.sub(r'\*(.+?)\*', r'<em>\1</em>', html)
    html = re.sub(r'`(.+?)`', r'<code>\1</code>', html)
    html = re.sub(r'!\[(.+?)\]\((.+?)\)', r'<img src="\2" alt="\1" />', html)
    html = html.replace('\n\n', '</p><p>')
    return f'<p>{html}</p>'


def create_draft_via_proxy(
    server_url: str,
    api_key: str,
    title: str,
    content: str,
    cover_url: str = "",
    cover_path: str = "",
    content_source_url: str = "",
) -> dict:
    """向代理服务器发送草稿创建请求。"""
    payload: dict = {
        "title": title,
        "content": content,
        "content_source_url": content_source_url,
    }

    if cover_path and os.path.exists(cover_path):
        with open(cover_path, "rb") as f:
            payload["cover_base64"] = base64.b64encode(f.read()).decode()
        payload["cover_filename"] = os.path.basename(cover_path)
        print(f"封面图（本地文件）: {cover_path}")
    elif cover_url:
        payload["cover_url"] = cover_url
        print(f"封面图（URL）: {cover_url}")

    endpoint = server_url.rstrip("/") + "/api/draft"
    resp = requests.post(
        endpoint,
        json=payload,
        headers={"X-API-Key": api_key, "Content-Type": "application/json"},
        timeout=120,
    )

    if resp.status_code == 401:
        raise Exception("API Key 无效，请检查 PROXY_API_KEY 配置")
    if resp.status_code != 200:
        raise Exception(f"服务器错误 {resp.status_code}: {resp.text[:200]}")

    return resp.json()


def main():
    parser = argparse.ArgumentParser(description="通过代理服务器创建微信公众号草稿")
    parser.add_argument("--server", default=os.getenv("WECHAT_PROXY_URL"), help="代理服务器地址（优先级最高，覆盖 .env）")
    parser.add_argument("--api-key", default=os.getenv("WECHAT_PROXY_API_KEY"), help="代理服务器 API Key（优先级最高，覆盖 .env）")
    parser.add_argument("--title", required=True, help="文章标题")
    parser.add_argument("--content", default="", help="文章内容（Markdown 或 HTML）")
    parser.add_argument("--file", default="", help="从文件读取文章内容（优先级高于 --content）")
    parser.add_argument("--html", action="store_true", help="内容已是排版好的 HTML，跳过 Markdown 转换")
    parser.add_argument("--theme", default="professional-clean", help="排版主题（--html 时忽略）")
    parser.add_argument("--cover-url", default="", help="封面图 URL")
    parser.add_argument("--cover-path", default="", help="封面图本地路径（优先级高于 URL）")
    parser.add_argument("--content-source-url", default="", help="原文链接")

    args = parser.parse_args()

    if not args.server:
        print("错误: 需要代理服务器地址（--server 或 .env / WECHAT_PROXY_URL）")
        print("提示: cd ~/.grok/skills/wechat-draft-proxy && cp .env.example .env 然后编辑填写")
        sys.exit(1)
    if not args.api_key:
        print("错误: 需要 API Key（--api-key 或 .env / WECHAT_PROXY_API_KEY）")
        print("提示: cd ~/.grok/skills/wechat-draft-proxy && cp .env.example .env 然后编辑填写")
        sys.exit(1)

    # 读取内容
    if args.file:
        with open(args.file, "r", encoding="utf-8") as f:
            raw_content = f.read()
    elif args.content:
        raw_content = args.content
    else:
        print("错误: 需要通过 --content 或 --file 提供文章内容")
        sys.exit(1)

    # Markdown → HTML（如需要）
    if args.html:
        html_content = raw_content
        print("已跳过 Markdown 转换（--html 模式）")
    else:
        print(f"转换 Markdown（主题：{args.theme}）...")
        html_content = markdown_to_html(raw_content, theme=args.theme)

    # 如果是完整 HTML 文件，提取 <body> 内容
    if "<body>" in html_content and "</body>" in html_content:
        import re as _re
        m = _re.search(r"<body>(.*?)</body>", html_content, _re.DOTALL)
        if m:
            html_content = m.group(1).strip()
            print("已从完整 HTML 中提取 <body> 内容")

    print(f"正在通过代理服务器创建草稿...")
    result = create_draft_via_proxy(
        server_url=args.server,
        api_key=args.api_key,
        title=args.title,
        content=html_content,
        cover_url=args.cover_url,
        cover_path=args.cover_path,
        content_source_url=args.content_source_url,
    )

    if result.get("success"):
        print(f"\n✅ 成功! 草稿已创建!")
        print(f"media_id: {result['media_id']}")
        if result.get("failed_images"):
            print(f"\n⚠️  {len(result['failed_images'])} 张图片处理失败:")
            for item in result["failed_images"]:
                print(f"  - {item['url']}: {item['error']}")
        print("\n请到微信公众号后台 -> 内容与互动 -> 草稿箱 查看")
    else:
        print(f"\n❌ 创建失败: {result.get('error', result)}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/wechat-draft-proxy/scripts/create_draft.js
#!/usr/bin/env node
/**
 * 微信公众号草稿箱创建工具（Node.js 代理服务器版）
 * 
 * 通过固定 IP 的代理服务器中转微信 API 调用，客户端不需要加入微信 IP 白名单。
 * 支持已排版 HTML 直接提交、封面图（本地/URL）、原文链接。
 * 
 * 推荐搭配 wechat-styler 先排版生成 HTML，再使用 --html 提交。
 * 
 * 用法:
 *   node create_draft.js --title "标题" --file article.html --html
 *   node create_draft.js --title "标题" --file article.md --cover-path cover.jpg
 * 
 * 配置优先级:
 *   1. 命令行 --server / --api-key
 *   2. 技能目录下的 .env 文件（cp .env.example .env）
 *   3. 环境变量 WECHAT_PROXY_URL / WECHAT_PROXY_API_KEY
 */

const fs = require('fs');
const path = require('path');

// ==================== .env 自动加载（零依赖，技能目录优先） ====================
// 仅当对应环境变量尚未设置时才从 .env 填充，避免覆盖 shell export
function loadEnvFile() {
  const envPath = path.join(__dirname, '../.env');
  if (!fs.existsSync(envPath)) return;
  try {
    const content = fs.readFileSync(envPath, 'utf8');
    for (const rawLine of content.split(/\r?\n/)) {
      const line = rawLine.trim();
      if (!line || line.startsWith('#')) continue;
      const eqIdx = line.indexOf('=');
      if (eqIdx === -1) continue;
      const key = line.slice(0, eqIdx).trim();
      let val = line.slice(eqIdx + 1).trim();
      // 去除首尾引号
      if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
        val = val.slice(1, -1);
      }
      if (key && !process.env[key]) {
        process.env[key] = val;
      }
    }
  } catch (e) {
    // 静默，后面会给出清晰的错误提示
  }
}

// 模块加载时立即尝试加载（CLI 直接运行最常见场景）
loadEnvFile();

// ==================== 参数解析 ====================

function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    server: process.env.WECHAT_PROXY_URL || '',
    apiKey: process.env.WECHAT_PROXY_API_KEY || '',
    title: '',
    content: '',
    file: '',
    html: false,
    theme: 'professional-clean', // 保留参数但 Node 版不使用专业排版
    coverUrl: '',
    coverPath: '',
    contentSourceUrl: '',
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    const next = args[i + 1];

    switch (arg) {
      case '--server':
        options.server = next || '';
        i++;
        break;
      case '--api-key':
        options.apiKey = next || '';
        i++;
        break;
      case '--title':
        options.title = next || '';
        i++;
        break;
      case '--content':
        options.content = next || '';
        i++;
        break;
      case '--file':
        options.file = next || '';
        i++;
        break;
      case '--html':
        options.html = true;
        break;
      case '--theme':
        options.theme = next || 'professional-clean';
        i++;
        break;
      case '--cover-url':
        options.coverUrl = next || '';
        i++;
        break;
      case '--cover-path':
        options.coverPath = next || '';
        i++;
        break;
      case '--content-source-url':
        options.contentSourceUrl = next || '';
        i++;
        break;
      case '-h':
      case '--help':
        printHelp();
        process.exit(0);
        break;
      default:
        if (arg.startsWith('-')) {
          console.error(`未知参数: ${arg}`);
          process.exit(1);
        }
    }
  }

  return options;
}

function printHelp() {
  console.log(`
微信公众号草稿创建工具 (Node.js 版)

用法:
  node create_draft.js --title "文章标题" --file article.html --html
  node create_draft.js --title "文章标题" --file content.md --cover-path cover.jpg

选项:
  --title <string>           文章标题（必填）
  --file <path>              从文件读取内容（优先级高于 --content）
  --content <string>         直接传入内容字符串
  --html                     内容已是排版好的 HTML，跳过 Markdown 转换
  --server <url>             代理服务器地址（最高优先，覆盖 .env）
  --api-key <key>            API Key（最高优先，覆盖 .env）
  --cover-path <path>        封面图本地路径
  --cover-url <url>          封面图远程 URL
  --content-source-url <url> 原文链接
  --theme <name>             排版主题（当前 Node 版仅记录，推荐先用 wechat-styler 处理）
  -h, --help                 显示帮助

配置:
  推荐做法（最顺滑）：
    cd ~/.grok/skills/wechat-draft-proxy
    cp .env.example .env
    # 编辑 .env 填入 WECHAT_PROXY_URL 和 WECHAT_PROXY_API_KEY

  命令行可覆盖：--server / --api-key
  环境变量也可使用（优先级低于 .env 和 CLI）

推荐流程:
  1. 使用 wechat-styler 排版生成 HTML
  2. node create_draft.js --title "xxx" --file styled.html --html
`);
}

// ==================== Markdown 转 HTML（内置简易转换） ====================

function markdownToHtml(markdown) {
  let html = markdown;

  // 1. 代码块（必须最先处理，避免内部被转义）
  html = html.replace(/```(\w*)\n?([\s\S]*?)```/g, (m, lang, code) => {
    return `\n<pre><code class="lang-${lang || 'text'}">${escapeHtml(code.trim())}</code></pre>\n`;
  });

  // 2. 行内代码
  html = html.replace(/`([^`]+)`/g, '<code>$1</code>');

  // 3. 图片和链接
  html = html.replace(/!\[([^\]]*)\]\(([^)]+)\)/g, '<img src="$2" alt="$1" />');
  html = html.replace(/\[([^\]]+)\]\(([^)]+)\)/g, '<a href="$2" target="_blank">$1</a>');

  // 4. 粗体 / 斜体
  html = html.replace(/\*\*([^*]+)\*\*/g, '<strong>$1</strong>');
  html = html.replace(/__([^_]+)__/g, '<strong>$1</strong>');
  html = html.replace(/(^|[^*])\*([^*]+)\*(?!\*)/g, '$1<em>$2</em>');
  html = html.replace(/(^|[^_])_([^_]+)_(?!_)/g, '$1<em>$2</em>');

  // 5. 标题
  for (let i = 6; i >= 1; i--) {
    const re = new RegExp(`^#{${i}}\\s+(.+)$`, 'gm');
    html = html.replace(re, `<h${i}>$1</h${i}>`);
  }

  // 6. 列表（逐行状态机，较可靠）
  const lines = html.split('\n');
  const out = [];
  let listStack = []; // [{type: 'ul'|'ol'}]

  const isUl = (l) => /^(?:-|\*)\s+(.+)$/.exec(l);
  const isOl = (l) => /^\d+\.\s+(.+)$/.exec(l);

  for (let rawLine of lines) {
    const line = rawLine.trimEnd();
    const ul = isUl(line);
    const ol = isOl(line);

    if (ul || ol) {
      const type = ul ? 'ul' : 'ol';
      const content = ul ? ul[1] : ol[1];

      // 关闭不匹配的列表
      while (listStack.length && listStack[listStack.length - 1].type !== type) {
        const t = listStack.pop().type;
        out.push(`</${t}>`);
      }
      // 打开新列表
      if (!listStack.length || listStack[listStack.length - 1].type !== type) {
        listStack.push({ type });
        out.push(`<${type}>`);
      }
      out.push(`<li>${content}</li>`);
    } else {
      // 非列表行：关闭所有打开的列表
      while (listStack.length) {
        const t = listStack.pop().type;
        out.push(`</${t}>`);
      }
      out.push(line);
    }
  }
  while (listStack.length) {
    const t = listStack.pop().type;
    out.push(`</${t}>`);
  }
  html = out.join('\n');

  // 7. 段落（简单处理连续空行）
  html = html.replace(/\n{3,}/g, '\n\n');           // 压缩多空行
  html = html.replace(/\n\n/g, '</p>\n<p>');

  // 8. 包裹裸文本为段落（跳过已知块级开头）
  const blockStart = /^\s*<(h[1-6]|ul|ol|pre|blockquote|img|table|hr|div|p|li|\/(ul|ol))/i;
  const wrappedLines = html.split('\n').map(line => {
    const t = line.trim();
    if (!t) return '';
    if (blockStart.test(t)) return line;
    return `<p>${t}</p>`;
  });
  html = wrappedLines.join('\n');

  // 9. 最终大力清理：把错误包在 p 里的块级元素"救"出来
  html = html.replace(/<p>\s*<(ul|ol|h[1-6]|pre)>/gi, '<$1>');
  html = html.replace(/<\/(ul|ol|h[1-6]|pre)>\s*<\/p>/gi, '</$1>');
  html = html.replace(/<p>\s*<\/(ul|ol)>/gi, '</$1>');
  html = html.replace(/<p><\/p>/g, '');
  // 列表之间多余 p
  html = html.replace(/<\/(ul|ol)>\s*<p>\s*<\/p>\s*<(ul|ol)>/gi, '</$1><$2>');

  return html.trim();
}

function escapeHtml(str) {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

// ==================== HTML 处理 ====================

function extractBodyContent(html) {
  if (html.includes('<body') && html.includes('</body>')) {
    const match = html.match(/<body[^>]*>([\s\S]*?)<\/body>/i);
    if (match) {
      console.log('已从完整 HTML 中提取 <body> 内容');
      return match[1].trim();
    }
  }
  return html;
}

// ==================== 代理服务器调用 ====================

async function createDraftViaProxy(serverUrl, apiKey, payload) {
  const endpoint = serverUrl.replace(/\/$/, '') + '/api/draft';

  const res = await fetch(endpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-API-Key': apiKey,
    },
    body: JSON.stringify(payload),
  });

  if (res.status === 401) {
    throw new Error('API Key 无效，请检查 PROXY_API_KEY 配置');
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`服务器错误 ${res.status}: ${text.slice(0, 200)}`);
  }

  return res.json();
}

// ==================== 主流程 ====================

async function main() {
  const args = parseArgs();

  if (!args.server) {
    console.error('错误: 需要代理服务器地址（--server 或 .env / WECHAT_PROXY_URL）');
    console.error('提示: 在 wechat-draft-proxy 目录执行 "cp .env.example .env" 然后编辑填写');
    process.exit(1);
  }
  if (!args.apiKey) {
    console.error('错误: 需要 API Key（--api-key 或 .env / WECHAT_PROXY_API_KEY）');
    console.error('提示: 在 wechat-draft-proxy 目录执行 "cp .env.example .env" 然后编辑填写');
    process.exit(1);
  }
  if (!args.title) {
    console.error('错误: --title 是必填参数');
    process.exit(1);
  }

  // 读取内容
  let rawContent = '';
  if (args.file) {
    if (!fs.existsSync(args.file)) {
      console.error(`错误: 文件不存在: ${args.file}`);
      process.exit(1);
    }
    rawContent = fs.readFileSync(args.file, 'utf-8');
  } else if (args.content) {
    rawContent = args.content;
  } else {
    console.error('错误: 需要通过 --content 或 --file 提供文章内容');
    process.exit(1);
  }

  // 转换或直接使用
  let htmlContent;
  if (args.html) {
    htmlContent = rawContent;
    console.log('已跳过 Markdown 转换（--html 模式）');
  } else {
    console.log(`转换 Markdown（内置简易转换器，推荐使用 wechat-styler + --html 获得专业排版）...`);
    htmlContent = markdownToHtml(rawContent);
  }

  htmlContent = extractBodyContent(htmlContent);

  // 构建请求 payload
  const payload = {
    title: args.title,
    content: htmlContent,
    content_source_url: args.contentSourceUrl || '',
  };

  // 封面图处理
  if (args.coverPath) {
    if (!fs.existsSync(args.coverPath)) {
      console.error(`错误: 封面图文件不存在: ${args.coverPath}`);
      process.exit(1);
    }
    const imageBuffer = fs.readFileSync(args.coverPath);
    payload.cover_base64 = imageBuffer.toString('base64');
    payload.cover_filename = path.basename(args.coverPath);
    console.log(`封面图（本地文件）: ${args.coverPath}`);
  } else if (args.coverUrl) {
    payload.cover_url = args.coverUrl;
    console.log(`封面图（URL）: ${args.coverUrl}`);
  }

  console.log('正在通过代理服务器创建草稿...');

  try {
    const result = await createDraftViaProxy(args.server, args.apiKey, payload);

    if (result.success) {
      console.log('\n✅ 成功! 草稿已创建!');
      console.log(`media_id: ${result.media_id}`);
      if (result.failed_images && result.failed_images.length > 0) {
        console.log(`\n⚠️  ${result.failed_images.length} 张图片处理失败:`);
        for (const item of result.failed_images) {
          console.log(`  - ${item.url}: ${item.error}`);
        }
      }
      console.log('\n请到微信公众号后台 -> 内容与互动 -> 草稿箱 查看');
    } else {
      console.error(`\n❌ 创建失败: ${result.error || JSON.stringify(result)}`);
      process.exit(1);
    }
  } catch (err) {
    console.error(`\n❌ 请求失败: ${err.message}`);
    process.exit(1);
  }
}

// 入口
if (require.main === module) {
  main().catch((e) => {
    console.error('未捕获错误:', e);
    process.exit(1);
  });
}

module.exports = { markdownToHtml, extractBodyContent, createDraftViaProxy };
